### PR TITLE
RORDEV-650 (WIP) Numeric string(s) array as indices rules argument won't pass decoder validation

### PR DESCRIPTION
**Pull request by sscarduzio**
Only tests added for now.

### CODE REVIEW

**Comment by sscarduzio** [comment]
I'm a bit stuck with this.  This is what I investigated so far:

- The Yaml parser seem to behave well, as you can see from the re-converted JSON.

image

- Following the debugger, the failure happens in the `IndicesRuleDecoder`
 
image

But I tragically get lost in the implicit typelevel stuff and cannot pinpoint what part of the decoding goes bad. Any advice?

**Comment by coutoPL** [comment]
I see that the problem is somewhere in the `JsonConfigStaticVariableResolver` class. Check `RawRorConfigBasedCoreFactory#createCoreFromRorSection` method - in `rorSection` object we have a list of strings. In `resolvedRorSection` there is a list of numbers.

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala [hunk 1]
```diff
@@ -54,17 +54,31 @@ object JsonConfigStaticVariableResolver {
...
   }
 
-  private def resolvedStringToJson(resolvedStr: String) = {
+  private def preserveNumbersAsStrings(original: Json, newValue: Json) = {
+    {
+      for {
+        _ <- original.asString
+        newNum <- newValue.asNumber
+      } yield Encoder.encodeString(newNum.toString)
+    }.getOrElse(newValue)
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala:73]: is it safe to "catch all exceptions" and return the `newValue` value? Maybe we should catch only a specific exception?

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala [hunk 2]
```diff
@@ -54,17 +54,31 @@ object JsonConfigStaticVariableResolver {
...
+          resolvedStringToJson(resolved, json)
+        else json
       }
   }
 
-  private def resolvedStringToJson(resolvedStr: String) = {
+  private def preserveNumbersAsStrings(original: Json, newValue: Json) = {
+    {
+      for {
+        _ <- original.asString
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala:70]: I don't know if I understand what is going on here. We try to get a string from JSON and ignore the result and then try to get a number from the new value and encode it as a string or if all of these fail we return the newValue. It's rather hacky, so we should do one of the following:
1. write more declarative or at least descriptive function
2. explain in comment the intention of the hack

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala [hunk 3]
```diff
@@ -54,17 +54,31 @@ object JsonConfigStaticVariableResolver {
...
         }
-        resolvedStringToJson(resolved)
+        if (resolved =!= str)
+          resolvedStringToJson(resolved, json)
+        else json
       }
   }
 
-  private def resolvedStringToJson(resolvedStr: String) = {
+  private def preserveNumbersAsStrings(original: Json, newValue: Json) = {
```
Comment (coutoPL) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala:67]: the method is used only by the `resolvedStringToJson` method, so it should be defined after it (for a sake of readability)

**Review by coutoPL — CHANGES_REQUESTED** [review]

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala (hunk 1 above)
Comment (sscarduzio) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala:73]: Wait, asString and asNumber return Option, so either they "manage" to parse as the respective type, or they return None. What is there for us to  catch here? 🤔

Comment (coutoPL) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala:73]: Ok, my bad

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala (hunk 2 above)
Comment (sscarduzio) [reply core/src/main/scala/tech/beshu/ror/accesscontrol/factory/JsonConfigStaticVariableResolver.scala]: Nevermind, I played type tetris a bit too long when I did this and I was brain dead. I simplified the algorithm now and it should be evident what's happening.

**Review by coutoPL — APPROVED** [review]
LGTM

**Comment by coutoPL** [comment]
@sscarduzio could you please add a changelog style description to the PR and paste the link to the PR in the corresponding jira?

**Pull request merged**